### PR TITLE
fix: Separate immediate response results from regular output.

### PR DESCRIPTION
**Pull request by dallen68**
Prevent callbacks after immediate response.

### CODE REVIEW

File: plugins/test/dynamic_test.cc [hunk 1]
```diff
@@ -407,68 +443,87 @@ void DynamicTest::CheckPhaseResults(const std::string& phase,
...
+    auto it = immediate_response.headers.find(header.key());
+    if (it != immediate_response.headers.end()) {
+      ADD_FAILURE() << absl::Substitute(
+          "[$0] Immediate response header '$1' value is '$2', expected removed",
+          phase, header.key(), it->second);
+    }
+  }
+  // Check serialized headers.
+  if (expected_response.headers_size() > 0) {
+    std::vector<std::string> headers_kv;
```
Comment (martijneken) [inline plugins/test/dynamic_test.cc:520]: ditto: How about `header_lines`

**Review by martijneken — COMMENTED** [review]
Looking good, just a few small comments

**Comment by mpwarres** [comment]
It sounds like you guys are mostly in agreement here, so don't let me stop you, but (just to offer another viewpoint) I'm not entirely convinced this change is needed.

AIUI a callback can have one of two outcomes: it can either mutate the existing request/response, or trigger an `ImmediateResponse`, which takes precedence (i.e. overrides) any "normal" output. The current fields of the `Expectation.Immediate` protobuf are needed because they are not already present in `Expectation`. But adding additional `has_header`, `no_header`, `headers` and `body` fields in `Expectation.Immediate` paralleling those already in `Expectation` seems redundant to me, since at the end of the day there can only be one response. I.e., there isn't a case where you would want to verify expectations on both the headers in an ImmediateResponse as well as normally mutated headers, since if an ImmediateResponse is present, you no longer care about the normally mutated headers.

File: plugins/test/runner.proto [hunk 2]
```diff
@@ -76,6 +75,14 @@ message Expectation {
     optional uint32 grpc_status = 2;
     // Reason for immediate response. Used for error or debug logging.
     optional string details = 3;
+    // Immediate response headers expected in the output.
+    repeated Header has_header = 5;
+    // Immediate response headers expected to be missing. Values are ignored.
+    repeated Header no_header = 6;
```
[the remark this answers is not in the document]
Comment (dallen68) [reply plugins/test/runner.proto:81]: This just lets tests set expectations for the headers within an immediate response the same way they would set expectations for a normal request/response. If you're suggesting that we only support string matching for immediate response headers we can do that, but it's not costing us anything to support has_header and no_header here.

File: plugins/test/framework.cc [hunk 3]
```diff
@@ -211,8 +215,9 @@ TestHttpContext::Result TestHttpContext::SendResponseHeaders(
 TestHttpContext::Result TestHttpContext::SendResponseBody(std::string body) {
   phase_logs_.clear();
   result_ = Result{};
-  body_buffer_.setOwned(std::move(body));
   current_callback_ = TestHttpContext::CallbackType::ResponseBody;
+  if (immediate_response_) return Result{};
```
[the remark this answers is not in the document]
Comment (dallen68) [reply plugins/test/framework.cc:219]: It doesn't matter right now, but it's probably better that we accurately set phase_, clear result_, and set current_callback_ before returning. Should we have a failure between callbacks it would be nice if phase_ accurately reported where we were instead of being out of date because we had an immediate response and never updated phase_.

File: plugins/test/framework.cc [hunk 4]
```diff
@@ -199,9 +201,11 @@ TestHttpContext::Result TestHttpContext::SendRequestBody(std::string body) {
 TestHttpContext::Result TestHttpContext::SendResponseHeaders(
     TestHttpContext::Headers headers) {
   phase_logs_.clear();
-  result_ = Result{.headers = std::move(headers)};
+  result_ = Result{};
   phase_ = proxy_wasm::WasmHeaderMapType::ResponseHeaders;
   current_callback_ = TestHttpContext::CallbackType::ResponseHeaders;
+  if (immediate_response_) return Result{};
```
[the remark this answers is not in the document]
Comment (dallen68) [reply plugins/test/framework.cc:207]: It doesn't matter right now, but it's probably better that we accurately set phase_, clear result_, and set current_callback_ before returning. Should we have a failure between callbacks it would be nice if phase_ accurately reported where we were instead of being out of date because we had an immediate response and never updated phase_.

File: plugins/test/framework.cc [hunk 5]
```diff
@@ -162,14 +162,15 @@ proxy_wasm::WasmResult TestHttpContext::sendLocalResponse(
       phase_ != proxy_wasm::WasmHeaderMapType::ResponseHeaders) {
     return proxy_wasm::WasmResult::BadArgument;
   }
-  result_.http_code = response_code;
-  result_.body = body_text;
-  result_.grpc_code = grpc_status;
-  result_.details = details;
-  result_.headers.clear();
+  result_.immediate_response = ImmediateResponse{};
```
[the remark this answers is not in the document]
Comment (dallen68) [reply plugins/test/framework.cc:165]: You can't do implicit type conversion from string_view to string within a struck constructor like this. I could set the other fields this way, but I figure it's probably best to just do it all one way.

**Comment by dallen68** [comment]
> It sounds like you guys are mostly in agreement here, so don't let me stop you, but (just to offer another viewpoint) I'm not entirely convinced this change is needed.
> 
> AIUI a callback can have one of two outcomes: it can either mutate the existing request/response, or trigger an `ImmediateResponse`, which takes precedence (i.e. overrides) any "normal" output. The current fields of the `Expectation.Immediate` protobuf are needed because they are not already present in `Expectation`. But adding additional `has_header`, `no_header`, `headers` and `body` fields in `Expectation.Immediate` paralleling those already in `Expectation` seems redundant to me, since at the end of the day there can only be one response. I.e., there isn't a case where you would want to verify expectations on both the headers in an ImmediateResponse as well as normally mutated headers, since if an ImmediateResponse is present, you no longer care about the normally mutated headers.

We could only check for immediate response headers or normal headers depending on the existence of an immediate response, but I think by having to explicitly say that you are checking for one or the other better conveys that these are in fact completely separate sets of headers.

**Comment by mpwarres** [comment]
> We could only check for immediate response headers or normal headers depending on the existence of an immediate response, but I think by having to explicitly say that you are checking for one or the other better conveys that these are in fact completely separate sets of headers.

Fair enough, SGTM

File: plugins/test/framework.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (martijneken) [reply plugins/test/framework.cc:166]: In that case you can just use the constructor:

`  .field = std::string(my_view),`

I won't insist on it.

File: plugins/test/runner.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (martijneken) [reply plugins/test/runner.proto:81]: If you want to keep it please refactor the header-checking logic you duplicated into `DynamicTest::CheckImmediateResponse`.

File: plugins/test/framework.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (martijneken) [reply plugins/test/framework.cc:220]: ACK.

File: plugins/test/framework.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (martijneken) [reply plugins/test/framework.cc:208]: ACK.

[the remark this answers is not in the document]
Comment (martijneken) [reply plugins/test/framework.cc:208]: Looking at it again I still think this should go first. The intent is to skip the whole callback (as if the request was actually terminated), so just skip all the local phase logic.

File: plugins/test/runner.proto [hunk 6]
```diff
@@ -59,13 +69,8 @@ message Input {
...
   // Headers expected in the output.
-  repeated Header has_header = 1;
-  // Headers expected to be missing. Values are ignored.
-  repeated Header no_header = 2;
-  // Generalized header matching against serialized "key: value" lines.
-  repeated StringMatcher headers = 3;
-
-  // Content accompanying an immediate response or body invocation result.
+  optional HeaderMatcher headers = 1;
+  // Content accompanying a body invocation result.
```
Comment (martijneken) [inline plugins/test/runner.proto:73]: I don't like that this unnecessarily introduces a breaking (backwards incompatible) proto change. People might already be using this API, let's not break it for no reason.

Also the most common test case now "stutters" with `headers { has_header: }` syntax. I get that it makes code factoring easier to have a reused HeaderMatcher proto, but it's not worth it IMO.

Mike's comment of "is this even necessary" rings true. All we really need to do in this PR is track an internal boolean saying "the plugin has sent immediate response" and then avoid firing subsequent callbacks.

**Comment by martijneken** [comment]
Discussed in person: there are 3 separate changes here, so let's split the PRs:
1. once immediate response is called, stop invoking more callbacks
2. separate the Results structs for headers + immediate response (internally) to avoid cross-pollination by plugins that set headers after calling immediate response (weird/bad edge case)
3. change the testing proto interface to explicitly test immediate response data (optional)

I think item 1 is the critical one, the others are nice to have.
